### PR TITLE
feat(Checkpoints): validate checkpoint identifiers

### PR DESCRIPTION
**Pull request by rickvdl**
## Description
Adds validation for checkpoint identifiers passed to the `.checkpoint()` method. Invalid identifiers are logged and dropped before listener callbacks, resolution, or UI work, returning a no-action result with `.invalidCheckpointIdentifier`. The behavior is identical between DEBUG and release.

As discussed the following validations are applied:
- Between 1 and 255 characters
- Start with an ASCII letter
- Contain only:
  - ASCII letters `A-Z`, `a-z`
  - Numbers `0-9`
  - Hyphens `-`
  - Underscores `_`
- No spaces, periods, or special characters

---

> [!NOTE]
> **Medium Risk**
> Changes runtime behavior for previously accepted invalid checkpoint names (no resolution instead of unknown/no-match), which could surprise integrators but avoids misconfigured dashboard IDs driving work.
> 
> **Overview**
> **Checkpoint identifiers** are now validated before resolution or UI. Invalid IDs still trigger `onCheckpointHit` and `onCheckpointCompleted`, but skip resolution and return a no-action result with **`invalidCheckpointIdentifier`** (logged at error level). Rules: 1–255 characters, must start with an ASCII letter, and only ASCII letters, digits, `_`, and `-`.
> 
> Adds **`CheckpointIdentifierValidator`** and documents the rules on the Swift and Objective-C `checkpoint` APIs. **`CustomVariableKeyValidator`** max key length is aligned from 256 to **255** characters (with updated debug log text). Tests cover the validator, manager paths, ObjC wrapping, and API surface.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 862b0b8f94123c13eaf65bbbf40478f636b62335. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by rickvdl** [comment]
@RCGitBot please test

File: Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        manager.listener = listener
+        let invalidIdentifiers = [
+            "",
+            "1checkout",
+            "_checkout",
+            "-checkout",
+            "check out",
+            "check.out",
+            "chéckout",
+            "a" + String(repeating: "1", count: 100)
```
Comment (ajpallares) [inline Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift]: Maybe add one with an emoji?

File: RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        return characters.dropFirst().allSatisfy(Self.isAllowedCharacter)
+    }
+
+    private static func isAllowedCharacter(_ character: UInt8) -> Bool {
+        return Self.isASCIILetter(character)
+            || (UInt8(ascii: "0")...UInt8(ascii: "9")).contains(character)
+            || character == UInt8(ascii: "_")
+            || character == UInt8(ascii: "-")
```
Comment (tonidero) [inline RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift:43]: So... once we add automatically tracked events like `$app_open`, `$app_foreground`,... We might need to skip this validation on the SDK side. I think it's fine for now, but just  something to remember.

**Review by tonidero — APPROVED** [review]
🚢

File: Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift
Comment (ajpallares) [inline Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift:1]: Shouldn't these tests use `CheckpointIdentifierValidator` directly? Seems like these tests are only indirectly checking the `CheckpointIdentifierValidator` through `CheckpointsManager`. I think we should split these:
1. The direct validator tests (you could even use @Test(arguments:) for each string to validate)
2. A couple of manager integration tests asserting the guard runs before the listener and resolver (those probably belong in `CheckpointsManagerTests.swift`)

File: RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+//  Created by Rick van der Linden.
+//
+
+enum CheckpointIdentifierValidator {
+
+    static func isValid(_ identifier: String) -> Bool {
+        let characters = identifier.utf8
+
+        guard
+            (1...100).contains(characters.count),
```
Comment (ajpallares) [inline RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift]: Should we create a `static let maxLength = 100` constant? We can reuse it in the error message

File: Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift
Comment (ajpallares) [inline Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift:1]: A few edge cases we could add:

- Leading/trailing whitespace (`" checkout"`, `"checkout "`, `"checkout\n"`). Probably the most likely mistake.
- Combined uppercase + `-` + `_` + digits (e.g. `"A-1_b"`) — each is tested alone but never together.
- The error message itself: it's new code and completely untested. Asserting `localizedDescription` contains the rejected identifier would be worth it, since that string is the only debugging aid the integrator gets today.
- The ObjC path under `ENABLE_CHECKPOINTS_OBJC`. `CheckpointsManagerTests` already has tests behind that flag, so an error-forwarding case should be possible?

File: RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift (hunk 3 above)
Comment (facumenzella) [inline RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift]: Where's does this comparison come from? 🤔

File: RevenueCatUI/Checkpoints/CheckpointsManager.swift [hunk 4]
```diff
@@ -76,6 +76,10 @@ final class CheckpointsManager {
         identifier: String,
         params: CheckpointParams
     ) async throws -> CheckpointResult {
+        guard CheckpointIdentifierValidator.isValid(identifier) else {
+            throw CheckpointError.invalidIdentifier(identifier)
+        }
```
Comment (ajpallares) [inline RevenueCatUI/Checkpoints/CheckpointsManager.swift:90]: I have some doubts about whether we want to throw here, because:
* Invalid custom variables don't throw. I know it's not the same case, but both are string supplied by the integrators, with two very different outcomes.
* There's already the `CheckpointNoActionReason.unknownCheckpoint`. This case feels very close to that one, so maybe a `CheckpointNoActionReason.invalidCheckpointIdentifier` would match closer I believe. 
* I'd say it's probably more common for developers to ignore a thrown error than a non-error outcome (e.g. `.invalidCheckpointIdentifier`).

WDYT?

Comment (facumenzella) [inline RevenueCatUI/Checkpoints/CheckpointsManager.swift:90]: Worth an error log I guess

**Review by ajpallares — COMMENTED** [review]
Looking good overall! I have some nits and some design questions

File: RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift (hunk 3 above)
Comment (facumenzella) [reply RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift]: shouldn't the 100 match khepri? there it only checks the identifier is [non-empty](https://github.com/RevenueCat/khepri/blob/d4fa46c3dbccd6a25235eaaaf8cc9ec6d25c18f8/khepri/api/developer_api_v2/domains/project_configuration/inputs/checkpoints.py#L31-L36) and the column is [`String(255)`](https://github.com/RevenueCat/khepri/blob/d4fa46c3dbccd6a25235eaaaf8cc9ec6d25c18f8/khepri/services/checkpoints/domain/model/checkpoint.py#L30-L33), so you could create a longer one through the API and never be able to call it. or is the 100 coming from somewhere else? (+1 on the constant btw)

File: Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+//  Created by Rick van der Linden.
+//
+
+@_spi(Internal) @testable import RevenueCat
+@_spi(CheckpointsInternal) @_spi(Internal) @testable import RevenueCatUI
+import XCTest
+
+@MainActor
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+final class CheckpointIdentifierValidatorTests: TestCase {
```
Comment (facumenzella) [inline Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift:21]: nit: nothing here touches CheckpointIdentifierValidator, it's all through the manager. Should these live in CheckpointsManagerTests ?

File: RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift (hunk 3 above)
Comment (ajpallares) [reply RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift]: Definitely, we should have this aligned everywhere. The 100 was a proposal from @rickvdl I believe (see [Slack thread](https://revenuecat.slack.com/archives/C0BKAEECUCW/p1787118058671999?thread_ts=1787058629.665559&cid=C0BKAEECUCW)), but it's open for discussion I guess. I do think these identifiers should not be too long, so 100 seems like a good compromise to me.

File: Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift
Comment (facumenzella) [reply Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift:1]: Worth sharing test cases with android :)

File: RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift (hunk 3 above)
Comment (rickvdl) [reply RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift]: Updated to match the final decision [here](https://revenuecat.slack.com/archives/C0BKAEECUCW/p1787135805546699?thread_ts=1787058629.665559&cid=C0BKAEECUCW)

File: RevenueCatUI/Checkpoints/CheckpointsManager.swift (hunk 4 above)
Comment (rickvdl) [reply RevenueCatUI/Checkpoints/CheckpointsManager.swift:90]: Added [here](https://github.com/RevenueCat/purchases-ios/blob/6581ae7c4241db4e35dd2c4aa49fee0d197c0290/RevenueCatUI/Checkpoints/CheckpointsManager.swift#L79-L85)

Comment (rickvdl) [reply RevenueCatUI/Checkpoints/CheckpointsManager.swift:90]: Great points, updated to do just that. Will now return `.noAction(.invalidCheckpointIdentifier)` and log the validation failure

File: Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift (hunk 1 above)
Comment (rickvdl) [reply Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift]: Done in https://github.com/RevenueCat/purchases-ios/blob/6581ae7c4241db4e35dd2c4aa49fee0d197c0290/Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift#L54

File: RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift (hunk 3 above)
Comment (rickvdl) [reply RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift]: Validations rules are now updated to the final decision taken [here](https://revenuecat.slack.com/archives/C0BKAEECUCW/p1787135805546699?thread_ts=1787058629.665559&cid=C0BKAEECUCW)

File: Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift
Comment (rickvdl) [reply Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift:1]: Added test cases for all of these (except the Objc one since the throwing error was removed).

File: Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift (hunk 5 above)
Comment (rickvdl) [reply Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift:21]: Good point, done in https://github.com/RevenueCat/purchases-ios/commit/6581ae7c4241db4e35dd2c4aa49fee0d197c0290

File: Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift
Comment (rickvdl) [reply Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift:1]: Great points, agree, that should both be covered in https://github.com/RevenueCat/purchases-ios/commit/6581ae7c4241db4e35dd2c4aa49fee0d197c0290

Comment (rickvdl) [reply Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift:1]: cc @tonidero for syncing the validations with Android. Will try to port them if I have time left today, if not could you please do?

Comment (tonidero) [reply Tests/RevenueCatUITests/Checkpoints/CheckpointIdentifierValidatorTests.swift:1]: I think @vegaro is actually going to handle this in Android 🙏 But definitely lmk if I should take over anything!

File: RevenueCatUI/Checkpoints/CheckpointsManager.swift [hunk 6]
```diff
@@ -76,6 +76,14 @@ final class CheckpointsManager {
         identifier: String,
         params: CheckpointParams
     ) async throws -> CheckpointResult {
+        guard CheckpointIdentifierValidator.isValid(identifier) else {
+            Logger.error(CheckpointIdentifierValidator.invalidIdentifierLogMessage(identifier))
```
Comment (ajpallares) [inline RevenueCatUI/Checkpoints/CheckpointsManager.swift:83]: It seems to me we should call the listener here passing this result?

File: RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+//      https://opensource.org/licenses/MIT
+//
+//  CheckpointIdentifierValidator.swift
+//
+//  Created by Rick van der Linden.
+//
+
+enum CheckpointIdentifierValidator {
+
+    static let maxLength = 255
```
Comment (ajpallares) [inline RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift:17]: Oh, Cursor noticed that this other validation check may be wrong?
https://github.com/RevenueCat/purchases-ios/blob/97c6ce01152d70dc2bdbf14db0622c52d99265c2/Sources/Misc/CustomVariableKeyValidator.swift#L41

It should be 1...255

**Review by ajpallares — COMMENTED** [review]
THanks for iterating on this! I have 2 suggestions

File: RevenueCatUI/Checkpoints/CheckpointsManager.swift (hunk 6 above)
Comment (rickvdl) [reply RevenueCatUI/Checkpoints/CheckpointsManager.swift:83]: Hmm yeah I think that would make sense as well, will update

File: RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift (hunk 7 above)
Comment (rickvdl) [reply RevenueCatUI/Checkpoints/CheckpointIdentifierValidator.swift:17]: Great catch 🤦 Updated

**Comment by rickvdl** [comment]
@RCGitBot please test
